src/utils/colorManipulation.ts

```diff
@@ -0,0 +1,208 @@
+  res[GREEN_CHANNEL_OFFSET] = Math.round(res[GREEN_CHANNEL_OFFSET] / numPixels);
+  res[BLUE_CHANNEL_OFFSET] = Math.round(res[BLUE_CHANNEL_OFFSET] / numPixels);
+  return res;
+}
+
+export function findVideoPositionOnCanvas(canvasDimensions: Dimensions, videoDimensions: Dimensions): Position {
+  const verticalPadding = (canvasDimensions.height - videoDimensions.height) / 2;
+  const horizontalPadding = (canvasDimensions.width - videoDimensions.width) / 2;
+  return {
+    top: verticalPadding,
+    right: canvasDimensions.width - horizontalPadding,
+    bottom: canvasDimensions.height - verticalPadding,
+    left: horizontalPadding
+  }
+}
+
+/**
+ * - A function that is used to smooth the color transition between divisions
+ * @param frame is the default canvas that holds the divison width/length colors]
+ * - Defines the boundaries for region convolution
+ * @param startRow
+ * @param startCol
+ * @param endRow
+ * @param endCol
+ * @param kernel_size is how large your kernel convolution size is (higher = laggier but smoother transition)
+ * @returns an ImageData
+ */
+
+export function regionConvolution(
+  frame: ImageData,
+  startRow: number,
+  startCol: number,
+  endRow: number,
+  endCol: number,
+  kernel_size: number
+): ImageData {
+  const frameCopy = new Uint8ClampedArray(frame.data.length);
+  let layers = Math.floor(kernel_size / 2);
+
+  for (let row = startRow; row < endRow; row++) {
+    for (let col = startCol; col < endCol; col++) {
+      const result = convolveRegion(
+        row,
+        col,
+        layers,
+        frame,
+        startRow,
+        startCol,
+        endRow,
+        endCol
+      );
+
+      let index = (row * frame.width + col) * 4;
+      
+      frameCopy[index] = result[0] / (kernel_size * kernel_size);
+      frameCopy[index + 1] = result[1] / (kernel_size * kernel_size);
+      frameCopy[index + 2] = result[2] / (kernel_size * kernel_size);
+      frameCopy[index + 3] = 255;
+    }
+  }
+  
+  return new ImageData(frameCopy, endRow - startRow, endCol - startCol); //i think that the height and width are off
+}
+
+export function convolveRegion(
+  row: number,
+  col: number,
+  layers: number,
+  frame: ImageData,
+  startRow: number,
+  startCol: number,
+  endRow: number,
+  endCol: number
+) {
+
+  let red = 0, blue = 0, green = 0, alpha = 255;
+  for (let kernel_row = row - layers; kernel_row < row + layers + 1; kernel_row++) {
+    for (let kernel_col = col - layers; kernel_col < col + layers + 1; kernel_col++) {
+      const rowOutOfBounds = kernel_row < startRow || kernel_row >= endRow;
+      const columnOutOfBounds = kernel_col < startCol || kernel_col >= endCol;
+      if (rowOutOfBounds || columnOutOfBounds) {
+        continue;
+      }
+
+      red += frame.data[(kernel_row * frame.width + kernel_col) * 4]
+      green += frame.data[((kernel_row * frame.width + kernel_col) * 4) +1]
+      blue += frame.data[((kernel_row * frame.width + kernel_col) * 4) +2]
```

Comment:
[nit] maybe worth saving the `(kernel_row * frame.width + kernel_col) * 4` value in a variable?

Reply:
replaced with index variable